### ISSUE TITLE
[PKG] C++ standard library

### ISSUE THREAD

**Issue by wongk** [google/prefab#127]
libc++_shared.so is constantly a problem. It gets included in every artifact that packages native libraries, which causes collisions all over the place. Ultimately we can exclude it from all of the artifacts, but it needs to be included somewhere. The question is where? If there were an artifact in Maven that included only this library, then each project can specify which version of the C++ library to include, and ultimately Maven will handle it for us. It also gives flexibility for app projects to specify exactly which version to include if different versions are present in the dependency graph.

**Comment by DanAlbert** [comment]
> It gets included in every artifact that packages native libraries, which causes collisions all over the place.

Can you give an example? This sounds like a bug.

**Comment by wongk** [comment]
Any time you build an android library with an external native build and shared C++ linkage, this .so file is included in the archive. If you have an app or a library that has multiple such dependencies you get collisions. `packagingOptions` is inflexible. If they are all using the same version of the NDK, you could simply `pickFirst`, but that may not be the case. Ultimately each project needs to be in control of the version of the C++ standard library that gets packaged.

**Comment by wongk** [comment]
Yeah I have read those docs but unfortunately they are not useful for us. At the end of the day we have many different modules with native components and a complex dependency graph. Static linking is exactly the opposite of what makes sense for us. I do agree with you that the prefab piece of this package is not what I am after with this request, and is not useful. It just so happens that providing this package would give me what I am asking for. If each individual component excludes `libc++_shared.so` from its packaging, then that enables downstream consumers to include the version that is most appropriate using maven.

**Comment by DanAlbert** [comment]
Without a more concrete picture of the problem you're trying to solve I don't know that I can offer any useful advice. I know that I won't have time to do this any time soon though (this is just one of many hats that I wear, and all of the others have deadlines I can't ignore), and it still really seems to me like this wouldn't be the _best_ fix, so if the better fix is actually an AGP feature or a change to some part of your workflow you actually stand a chance of getting a fix in a timeframe that matters to you.

**Comment by wongk** [comment]
I suspect the _best_ fix would be to improve `packagingOptions` to support these kinds of use cases. More specifically, to allow the build script author to specify that a particular file (libc++_shared.so in this case) should be used from a particular Maven package (or from the current project).

**Comment by DanAlbert** [comment]
Makes sense to me. File the AGP bug. If it's still stuck in triage after a few (working) days lmk and I'll try to get it to the right people.
